File: 240_search_2d_matrix_ii/binary_search.py
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        """
        Complexity: O(mlgn)
        Space: O(1)
        """
        def binary_search(row, start, end, t):
            # can modify it with a while loop instead of recursion
            mid = (end- start + 1)//2
            if row[start + mid] == t:
                return True
            elif start == end:
                return False
            elif row[start + mid] > t:
                return binary_search(row, start, start+mid-1, t)
            else:
                return binary_search(row, start+mid, end, t)
            
        if matrix == [] or matrix[0] == []:
            return False
        m = len(matrix)
        n = len(matrix[0])
        for i in range(m):
            if target >= matrix[i][0] and target <= matrix[i][-1]:
                if binary_search(matrix[i], 0, n-1, target):
                    return True
            else:
                continue
        return False

File: 240_search_2d_matrix_ii/binary_search.cpp
class Solution {
public:
    bool binary_search(vector<int>& row, int start, int end, int target){
        while(start <= end){
            int mid = (end - start + 1) / 2;
            if(row[start + mid] == target) return true;
            else if(start == end) return false;
            if(row[start + mid] > target) end = start+mid-1;
            else start = start+mid;
        }
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(matrix.size() == 0 || matrix[0].size() == 0) return false;
        int n_col = matrix[0].size();
        for(int i = 0; i < matrix.size(); ++i){
            if(matrix[i][0] <= target && matrix[i][n_col-1] >= target){
                if(binary_search(matrix[i], 0, n_col-1, target)) return true;
            }
            else continue;
        }
        return false;
    }
};
